src/Compilers/CSharp/Test/Semantic/Semantics/PatternMatchingTests4.cs

```diff
@@ -921,5 +921,468 @@ class C1
+        {
+            var source = @"using System;
+class Program
+{
+    static void Main()
+    {
+        object o = new _();
+        foreach (var e in new[] { null, o, null })
+        {
+            Console.Write(e is _);
+        }
+    }
+}
+class _
+{
+}";
+            var compilation = CreatePatternCompilation(source);
+            compilation.VerifyDiagnostics(
+                // (9,32): warning CS8513: The name '_' refers to the type '_', not the discard pattern. Use '@_' for the type, or 'var _' to discard.
+                //             Console.Write(e is _);
+                Diagnostic(ErrorCode.WRN_IsTypeNamedUnderscore, "_").WithArguments("_").WithLocation(9, 32)
+                );
+            CompileAndVerify(compilation, expectedOutput: "FalseTrueFalse");
+        }
+
+        [Fact]
+        public void DiscardVsTypeInIs_02()
+        {
+            var source = @"using System;
+class Program
+{
+    static void Main()
+    {
+        object o = new _();
+        foreach (var e in new[] { null, o, null })
+        {
+            Console.Write(e is _ x);
+        }
+    }
+}
+class _
+{
+}";
+            var compilation = CreatePatternCompilation(source);
+            compilation.VerifyDiagnostics(
+                // (9,32): warning CS8513: The name '_' refers to the type '_', not the discard pattern. Use '@_' for the type, or 'var _' to discard.
+                //             Console.Write(e is _ x);
+                Diagnostic(ErrorCode.WRN_IsTypeNamedUnderscore, "_").WithArguments("_").WithLocation(9, 32),
+                // (9,34): error CS1003: Syntax error, ',' expected
+                //             Console.Write(e is _ x);
+                Diagnostic(ErrorCode.ERR_SyntaxError, "x").WithArguments(",", "").WithLocation(9, 34),
+                // (9,34): error CS0103: The name 'x' does not exist in the current context
+                //             Console.Write(e is _ x);
+                Diagnostic(ErrorCode.ERR_NameNotInContext, "x").WithArguments("x").WithLocation(9, 34)
+                );
+        }
+
+        [Fact]
+        public void DiscardVsTypeInIs_03()
+        {
+            var source = @"using System;
+class Program
+{
+    static void Main()
+    {
+        object o = new _();
+        foreach (var e in new[] { null, o, null })
+        {
+            Console.Write(e is var _);
+        }
+    }
+}
+class _
+{
+}";
+            var compilation = CreatePatternCompilation(source);
+            compilation.VerifyDiagnostics(
+                );
+            CompileAndVerify(compilation, expectedOutput: "TrueTrueTrue");
+        }
+
+        [Fact]
+        public void DiscardVsTypeInIs_04()
+        {
+            var source = @"using System;
+class Program
+{
+    static void Main()
+    {
+        object o = new _();
+        foreach (var e in new[] { null, o, null })
+        {
+            if (e is @_)
+            {
+                Console.Write(""3"");
+            }
```

Comment:
Consider using `Console.Write(e is @_);` to match the tests above. Same comment for tests below.

Reply:
This is intentional to simplify expected output, though in many of these tests there is no observable difference.

---
In reply to: [230176393](https://github.com/dotnet/roslyn/pull/30847#discussion_r230176393) [](ancestors = 230176393)